src/fields.jl

```diff
@@ -282,17 +340,51 @@ function ∂x!(output::Field2D{Centre, Centre}, input::Field2D{Face, Centre, Gri
         output.data[nx, j] = (input.data[1, j] - input.data[nx, j]) / dx
     end
 
-    for j in 1:ny, i = 1:nx-1
+    for j in 1:ny, i = 1:nx
         output.data[i, j] = δxᶜᶜ(i, j, input) / dx
     end
+
+    fill_halos!(output)
     
     return nothing
 end
 
+"""
+    ∂x!(output::Field2D{face, Centre}, input::Field2D{Face, Centre, Grid2D{Periodic, Periodic}})
+
+Compute the derivative of a 2D `input` field onto the location where the `output` field lives
+for 2D grids with periodic boundary conditions.
+
+Compute the derivative of a 2D field of `data` from face to face in x-direction.
+"""
+function ∂x!(output::Field2D{Face, Centre}, input::Field2D{Face, Centre, Grid2D{Periodic, Periodic}})
+    nx, ny = input.grid.nx, input.grid.ny
+    dx = input.grid.dx
+        
+    for j in 1:ny
+        output.data[nx, j] = (input.data[1, j] - input.data[nx, j]) / dx
+    end
+
+    for j in 1:ny, i = 1:nx
+        output.data[i, j] = δxᶜᶜ(i, j, input) / dx
+    end
+
+    fill_halos!(output)
+
+    𝐼x!(output, input)
```

Comment:
I don't understand this step!

Reply:
> There are lot's of leftovers from the pre-halo era. Also there are a few points where you did changes that I don't understand why... seems to me that those changes should break the code but the tests pass which is alarming...
> 
> Also, note that I did make some annotations but that doesn't mean that I marked all the places that I feel there are left over code from the pre-halo era!

Yes there are. I forgot to change them, I was coding in half sleep. The tests work because by using halo points, we are using the same thing as taking the first and last points to estimate values at boundary points. I don't think I can explain it via chat.